In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# creating a path to load data fames to python 
path = r"C:\Users\pshan\Downloads\InstaCart Basket Analysis"

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [6]:
df_test = pd.DataFrame()  #creates a new dataframe called df_test. and then putting mixed type values to the column mix
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.dtypes

mix    object
dtype: object

In [8]:
# check for mix data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# updating datatype to str
df_test['mix'] = df_test['mix'].astype('str')

In [48]:
# running stats on df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [50]:
# finding median of prices column
df_prods['prices'].median()

7.1

In [51]:
# finding mean of prices column
df_prods['prices'].mean()

9.994135995009357

In [58]:
# finding outliers in prices column
outliers = df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

In [61]:
outliers

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

## handling outliers
#### because of the 2 extreme outliers the mean value is affected, Replacing the outliers values in the prices column with the meadian value of 7.1. makes more sense

In [62]:
 # replacing the outliers with median value of 7.1
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [69]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

### Observation 
#### After conducting exploratory analysis on the df_products DataFrame, outliers were identified in the prices column. The column had a mean of 9.99 and a standard deviation of 453.6. This unusually high standard deviation, more than 2 standard deviations away from the mean, highlighted the presence of outlier values.
#### Specifically, two entries were identified as outliers: prices of 14,900 and 99,999. These outliers were replaced with the median value of 7.1. Replacing these extreme values with the median is recommended because the minimum value of 1 and the maximum value of 99,999 create a wide range, making the median a more robust measure compared to the mean, which is sensitive to outliers.
#### After replacing the outliers, the mean of the prices column decreased to 7.6, reflecting a more accurate representation of the data.

In [65]:
# finding missing values in products df
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [66]:
# creating new data frame containing only null values in product name
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [67]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [70]:
# creating a new clean data fram that conatins no null values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [71]:
df_prods_clean.shape

(49677, 5)

In [72]:
df_prods.shape

(49693, 5)

In [73]:
# finding duplicates in the data frame
df_dup = df_prods_clean[df_prods_clean.duplicated()]

In [74]:
df_dup

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [75]:
# creating a new data frame without duplicates
df_prods_clean_no_dup =  df_prods_clean.drop_duplicates()

In [76]:
df_prods_clean_no_dup.shape

(49672, 5)

# df_ords

In [32]:
# step 2:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### After examining the values in each column, no unreasonable values were found

In [33]:
# step 3: 
# checking for mixed type in orders data set
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)


####  step 4:
#### The `df_ords` DataFrame contains no mixed-type data in any of its columns or variables. This was verified by using a function to search for mixed-type data within the columns, which returned a result of NULL.

In [34]:
# step 5: checking for missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [41]:
df_ords[df_ords['days_since_prior_order'].isnull() == True]

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

#### step 6: 
#### after examinig the data it was found that, The `days_since_prior_order` column in the original `df_ords` DataFrame contains a total of 206,209 missing values. Each unique `user_id` has exactly one missing value in this column, which corresponds to their very first order. This is logical, as there is no prior order for the first purchase of a user. Therefore, it is unnecessary to remove these missing values.

In [45]:
# step 7:
# checking for duplicates in the df_ords
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

 #### step 8: 
 #### there was found to be no duplicates in the entire df_ords data frame

In [80]:
# step 9:
# exporting the data sets
df_prods_clean_no_dup.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

In [81]:
# exporting the ords data sets
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index = False)